source/fuzz/transformation_inline_function.cpp

```diff
@@ -0,0 +1,176 @@
+      cloned_label_instruction->SetResultId(fresh_id);
+      function_call_instruction->InsertBefore(
+          std::unique_ptr<opt::Instruction>(cloned_label_instruction));
+      fuzzerutil::UpdateModuleIdBound(ir_context, fresh_id);
+    }
+
+    for (auto& instruction : block) {
+      // The function return instruction will not be inlined.
+      if (&instruction == &*called_function->tail()->tail()) {
+        continue;
+      }
+
+      // Replaces the operand ids with their mapped result ids.
+      auto cloned_instruction = instruction.Clone(ir_context);
+      cloned_instruction->ForEachInId(
+          [this, called_function, function_call_instruction](uint32_t* id) {
+            // If the id is mapped, then set it to its mapped value.
+            if (message_.result_id_map().count(*id)) {
+              *id = message_.result_id_map().at(*id);
+              return;
+            }
+
+            uint32_t parameter_index = 0;
+            called_function->ForEachParam(
+                [id, function_call_instruction,
+                 &parameter_index](opt::Instruction* parameter_instruction) {
+                  // If the id is a function parameter, then set it to the
+                  // parameter value passed in the function call instruction.
+                  if (*id == parameter_instruction->result_id()) {
+                    *id = function_call_instruction->GetSingleWordInOperand(
+                        parameter_index + 1);
+                  }
+                  parameter_index++;
+                });
+          });
+
+      // If |cloned_instruction| has a result id, then set it to its mapped
+      // value.
+      if (cloned_instruction->HasResultId() &&
+          message_.result_id_map().count(cloned_instruction->result_id())) {
+        uint32_t result_id =
+            message_.result_id_map().at(cloned_instruction->result_id());
+        cloned_instruction->SetResultId(result_id);
+        fuzzerutil::UpdateModuleIdBound(ir_context, result_id);
+      }
+
+      // If the called function return instruction is OpReturnValue and the
+      // |cloned_instruction| result id is the returned value, then sets the
+      // |cloned_instruction| result id to the function call result id.
+      if (cloned_instruction->result_id() ==
+          called_function->GetReturnValueId()) {
+        cloned_instruction->SetResultId(function_call_instruction->result_id());
+      }
+
+      if (cloned_instruction->opcode() == SpvOpVariable) {
+        // All OpVariable instructions in a function must be in the first block
+        // in the function.
+        caller_function->begin()->begin()->InsertBefore(
+            std::unique_ptr<opt::Instruction>(cloned_instruction));
+      } else {
+        function_call_instruction->InsertBefore(
+            std::unique_ptr<opt::Instruction>(cloned_instruction));
+      }
+    }
+  }
+
+  ir_context->InvalidateAnalysesExceptFor(opt::IRContext::kAnalysisNone);
+}
```

Comment:
What happens to function_call_instruction - does it get removed from the module?  (It should right?  I didn't see the code to do it, though.)

From your tests, the instruction appears to be gone, but I don't see how this happens.

Oh - in the test you kill the instruction!  That should be part of the transformation.

Reply:
Done.